core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1716,7 +1739,31 @@ private[spark] class DAGScheduler(
             if (runningStages.contains(shuffleStage) && shuffleStage.pendingPartitions.isEmpty) {
               if (!shuffleStage.shuffleDep.shuffleMergeFinalized &&
                 shuffleStage.shuffleDep.getMergerLocs.nonEmpty) {
-                scheduleShuffleMergeFinalize(shuffleStage)
+                // Check if a finalize task has already been scheduled. This is to prevent the
+                // following scenario: Stage A attempt 0 fails and gets retried. Stage A attempt 1
+                // succeeded, triggering the scheduling of shuffle merge finalization. However,
+                // tasks from Stage A attempt 0 might still be running and sending task completion
+                // events to DAGScheduler. This check prevents multiple attempts to schedule merge
+                // finalization get triggered due to this.
+                if (shuffleStage.shuffleDep.getFinalizeTask.isEmpty) {
+                  // If total shuffle size is smaller than the threshold, attempt to immediately
+                  // schedule shuffle merge finalization and process map stage completion.
+                  val totalSize = Try(mapOutputTracker
+                    .getStatistics(shuffleStage.shuffleDep).bytesByPartitionId.sum).getOrElse(0L)
+                  if (totalSize < shuffleMergeWaitMinSizeThreshold) {
+                    // Check if we can process map stage completion. If shuffle merge finalization
+                    // is already triggered because push completion ratio was reached earlier,
+                    // we cannot process map stage completion, but have to wait for the finalization
+                    // to finish. This is because it's not straightforward to interrupt the
+                    // finalization task and undo what it might have already done.
+                    if (scheduleShuffleMergeFinalize(shuffleStage, delay = 0,
+                      registerMergeResults = false)) {
```

Comment:
I am not sure of the semantics of this code path are ideal.

There are two cases:
a) We dont want to wait for executors to finish push, but want to trigger finalization at ESS immediately - so that any pushed data already available is finalized.
b) We dont care about pushed data, discard push entirely - move on the next stage.

We are doing (b) when `totalSize` < `shuffleMergeWaitMinSizeThreshold`.

While waiting for `shuffleMergeResultsTimeoutSec` for (a) might not be optimal for quick shuffles (since there can be small and large shuffles in same app), allowing some time for ESS to complete and leveraging whatever subset of ESS which could respond back to driver in time (and not discarding all the merge result entirely) can be useful (even if timeout is aggressively small).

Thoughts ?

Reply:
Also +CC @Ngone51 on this.

Reply:
This comment is subsumed by more details [here](https://github.com/apache/spark/pull/33896#discussion_r742415348).
Resolving this conversation.